homeassistant/components/guardian/__init__.py

```diff
@@ -89,33 +125,115 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        self._api_lock = api_lock
+        self._client = client
+        self._entry = entry
+        self._hass = hass
+        self._listeners = []
+        self._paired_uids = set()
+
+    async def async_pair_sensor(self, uid: str) -> None:
+        """Add a new paired sensor coordinator."""
+        LOGGER.info("Adding paired sensor: %s", uid)
+
+        self._paired_uids.add(uid)
+
+        coordinator = self._hass.data[DOMAIN][DATA_COORDINATOR][self._entry.entry_id][
+            API_SENSOR_PAIRED_SENSOR_STATUS
+        ][uid] = GuardianDataUpdateCoordinator(
+            self._hass,
+            client=self._client,
+            api_name=f"{API_SENSOR_PAIRED_SENSOR_STATUS}_{uid}",
+            api_coro=lambda: self._client.sensor.paired_sensor_status(uid),
+            api_lock=self._api_lock,
+            valve_controller_uid=self._entry.data[CONF_UID],
+        )
+        await coordinator.async_request_refresh()
+
+        async_dispatcher_send(
+            self._hass,
+            SIGNAL_PAIRED_SENSOR_COORDINATOR_ADDED.format(self._entry.data[CONF_UID]),
+            uid,
+        )
+
+    async def async_process_latest_paired_sensor_uids(self) -> None:
+        """Process a list of new UIDs."""
+        try:
+            uids = set(
+                self._hass.data[DOMAIN][DATA_COORDINATOR][self._entry.entry_id][
+                    API_SENSOR_PAIR_DUMP
+                ].data["paired_uids"]
+            )
+        except KeyError:
+            # Sometimes the paired_uids key can fail to exist; the user can't do anything
+            # about it, so in this case, we quietly abort and return:
+            return
+
+        if uids == self._paired_uids:
+            return
+
+        old = self._paired_uids
+        new = self._paired_uids = set(uids)
+
+        tasks = [self.async_pair_sensor(uid) for uid in new.difference(old)]
+        tasks += [self.async_unpair_sensor(uid) for uid in old.difference(new)]
+
+        if tasks:
+            await asyncio.gather(*tasks)
+
+    async def async_unpair_sensor(self, uid: str) -> None:
+        """Remove a paired sensor coordinator."""
+        LOGGER.info("Removing paired sensor: %s", uid)
+
+        # Clear out objects related to this paired sensor:
+        self._paired_uids.remove(uid)
+        self._hass.data[DOMAIN][DATA_COORDINATOR][self._entry.entry_id][
+            API_SENSOR_PAIRED_SENSOR_STATUS
+        ].pop(uid)
+
+        # Remove the paired sensor device from the device registry (which will
+        # clean up entities and the entity registry):
+        dev_reg = await self._hass.helpers.device_registry.async_get_registry()
+        device = dev_reg.async_get_or_create(
+            config_entry_id=self._entry.entry_id, identifiers={(DOMAIN, uid)}
+        )
+        dev_reg.async_remove_device(device.id)
+
+
+class GuardianEntity(CoordinatorEntity):
+    """Define a base Guardian entity."""
+
+    def __init__(  # pylint: disable=super-init-not-called
```

Comment:
Don't disable pylint. Either call super init or don't inherit from CoordinatorEntity.

Reply:
This isn't straightforward, as both `PairedSensorEntity` and `ValveControllerEntity` need some of what `CoordinatorEntity` offers, but `ValveControllerEntity` has multiple controllers and not one. I'll open up a new PR addressing the other feedback and we can discuss there.